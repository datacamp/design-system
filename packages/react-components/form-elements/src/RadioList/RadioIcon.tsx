import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css } from '@emotion/core';
import React from 'react';

const RadioIcon: React.FC<{
  checked: boolean;
  disabled: boolean;
  error: boolean;
}> = ({ checked, disabled, error }) => {
  let color = tokens.color.neutral.beige400.value.hex;
  if (checked) color = tokens.color.primary.blue.value.hex;
  if (error) color = tokens.color.primary.red.value.hex;

  return (
    <>
      <span
        css={css({
          borderColor: color,
          borderRadius: '50%',
          borderStyle: 'solid',
          borderWidth: '2px',
          boxSizing: 'border-box',
          height: '18px',
          opacity: disabled ? 0.3 : 1,
          position: 'absolute',
          top: '0px',
          width: '18px',
        })}
      />
      {checked && (
        <span
          css={{
            backgroundColor: color,
            borderRadius: '50%',
            boxSizing: 'border-box',
            height: '8px',
            marginLeft: '5px',
            opacity: disabled ? 0.3 : 1,
            position: 'absolute',
            top: '5px',
            width: '8px',
          }}
        />
      )}
    </>
  );
};

export default RadioIcon;
