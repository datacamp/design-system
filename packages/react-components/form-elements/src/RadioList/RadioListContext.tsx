import React from 'react';

export interface RadioListContextInterface {
  disabled: boolean;
  hasError: boolean;
  name: string;
  onChange: (value: string) => void;
  value: string;
}

const RadioListContext = React.createContext<RadioListContextInterface | null>(
  null,
);

export default RadioListContext;
