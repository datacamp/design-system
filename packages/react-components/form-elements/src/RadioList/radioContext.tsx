import React from 'react';

export interface RadioContextInterface {
  className?: string;
  label?: string;
  name: string;
  onChange: (value: string) => void;
  parsedDataAttributes: { [key: string]: string };
  value: string;
}
const RadioContext = React.createContext<RadioContextInterface | null>(null);

export const RadioContextProvider = RadioContext.Provider;
export const RadioContextConsumer = RadioContext.Consumer;
export default RadioContext;
