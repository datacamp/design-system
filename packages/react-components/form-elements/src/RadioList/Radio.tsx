import { Text } from '@datacamp/waffles-text';
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css, SerializedStyles } from '@emotion/core';
import React from 'react';

import { RadioContextConsumer } from './radioContext';

export interface RadioProps {
  /**
   * It blocks user interaction.
   */
  disabled?: boolean;
  /**
   * The value of the radio list. This should be controlled by listening to onChange.
   */
  value: string;
}

const iconStyle: SerializedStyles = css({
  ':active': { boxShadow: 'inset 0 0 0 2px currentColor' },
  borderRadius: '50%',
  boxShadow: 'inset 0 0 0 1px currentColor',
  height: '16px',
  marginTop: '2px',
  position: 'absolute',
  width: '16px',
});

// adds a filled circle inside the radio button if it's checked
const checkedIconStyle = css({
  borderRadius: '50%',
  height: '8px',
  marginLeft: '4px',
  marginTop: '6px',
  position: 'absolute',
  width: '8px',
});

const textStyle = css({
  marginLeft: tokens.size.space[24].value,
});

const divStyle = css({
  ':focus, :active': {
    color: tokens.color.opaque.primary.value.hex,
  },
  color: tokens.color.opaque.greyLight.value.hex,
  display: 'flex',
  marginTop: '14px',
  position: 'relative',
});

const errorStyle = css({
  ':active span:first-of-type': {
    boxShadox: `inset 0 0 0 2px ${tokens.color.opaque.red.value.hex}`,
  },
  ':focus-within span:first-of-type': {
    boxShadow: `inset 0 0 0 2px ${tokens.color.opaque.red.value.hex}`,
  },
  color: tokens.color.opaque.red.value.hex,
});

const Radio = ({ disabled, value }: RadioProps): React.ReactElement => {
  const isDisabled = css({
    boxShadow: disabled
      ? `inset 0 0 0 1px ${tokens.color.opaque.greyLight.value.hex}`
      : `inset 0 0 0 2px ${tokens.color.opaque.primary.value.hex}`,
  });
  return (
    <RadioContextConsumer>
      {contextValue =>
        contextValue && (
          <div
            css={css(
              divStyle,
              {
                ':active span:first-of-type': contextValue.errorMessage
                  ? {
                      boxShadow: `inset 0 0 0 2px ${tokens.color.opaque.red.value.hex}`,
                    }
                  : isDisabled,
                ':focus-within span:first-of-type': contextValue.errorMessage
                  ? {
                      boxShadow: `inset 0 0 0 2px ${tokens.color.opaque.red.value.hex}`,
                    }
                  : isDisabled,
              },
              contextValue.errorMessage && !disabled && errorStyle
            )}
          >
            <label
              css={{
                'input[type="radio"] ': {
                  opacity: 0,
                  position: 'absolute',
                  width: 0,
                },
              }}
              htmlFor={value}
            >
              <input
                checked={contextValue.value === value}
                disabled={disabled}
                id={value}
                name={contextValue.name}
                onChange={() => {
                  contextValue.onChange(value);
                }}
                type="radio"
                {...contextValue.parsedDataAttributes}
              />
              <span
                css={
                  contextValue.value === value
                    ? css(
                        iconStyle,
                        {
                          boxShadow: disabled
                            ? `inset 0 0 0 1px ${tokens.color.opaque.greyLight.value.hex}`
                            : `inset 0 0 0 1px ${tokens.color.opaque.primary.value.hex}`,
                        },
                        contextValue.errorMessage && {
                          ':active': {
                            boxShadow: `inset 0 0 0 2px ${tokens.color.opaque.red.value.hex}`,
                          },
                          ':focus': {
                            boxShadow: `inset 0 0 0 2px ${tokens.color.opaque.red.value.hex}`,
                          },
                          boxShadow: `inset 0 0 0 1px ${tokens.color.opaque.red.value.hex}`,
                        }
                      )
                    : css(iconStyle)
                }
              />
              {contextValue.value === value && (
                <span
                  css={css(
                    checkedIconStyle,
                    {
                      backgroundColor: disabled
                        ? tokens.color.opaque.greyLight.value.hex
                        : tokens.color.opaque.primary.value.hex,
                    },
                    contextValue.errorMessage && {
                      backgroundColor: tokens.color.opaque.red.value.hex,
                    }
                  )}
                />
              )}

              <Text
                css={
                  disabled
                    ? css(textStyle, {
                        color: tokens.color.opaque.greyLight.value.hex,
                      })
                    : textStyle
                }
              >
                {value}
              </Text>
            </label>
          </div>
        )
      }
    </RadioContextConsumer>
  );
};

export default Radio;
