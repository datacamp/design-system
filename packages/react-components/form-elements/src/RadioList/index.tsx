import { childrenOfType } from 'airbnb-prop-types';
import PropTypes from 'prop-types';
import React, { ReactElement } from 'react';

import Label from '../Label';
import Radio, { RadioProps } from './Radio';
import RadioListContext from './RadioListContext';

export interface RadioListProps {
  children: ReactElement<RadioProps>[];
  /**
   * When provided the description will show underneath the label. Only
   * available when 'label' is also specified.
   */
  description?: string;
  /**
   * Disables the whole RadioList when true.
   */
  disabled?: boolean;
  /**
   * When provided the errorMessage will show underneath the RadioList.
   */
  errorMessage?: string;
  /**
   * Renders label text above the RadioList.
   */
  label: string;
  /**
   * Sets the name attribute on the rendered Radio buttons.
   */
  name: string;
  /**
   * Called when the user requests a change to the value of the RadioList. This
   * should be used to set the value.
   */
  onChange: (value: string) => void;
  /**
   * It defines whether the RadioList is required or not. If required=true it
   * adds the text 'Required' on the top-right of the RadioList, if
   * required=false it adds 'Optional'. The default value is undefined, which
   * doesn't add anything.
   */
  required?: boolean;
  /**
   * The current value of the RadioList. This should be controlled by listening to
   * onChange.
   */
  value: string;
}

const RadioList = ({
  children,
  description,
  disabled = false,
  errorMessage,
  label,
  name,
  onChange,
  required,
  value,
}: RadioListProps): ReactElement => (
  <Label
    as="div"
    description={description}
    errorMessage={errorMessage}
    label={label}
    required={required}
  >
    <div>
      <RadioListContext.Provider
        value={{
          disabled,
          hasError: !!errorMessage,
          name,
          onChange,
          value,
        }}
      >
        {children}
      </RadioListContext.Provider>
    </div>
  </Label>
);

RadioList.propTypes = {
  /**
   * Takes multiple `Radio` components as children.
   */
  children: PropTypes.arrayOf(childrenOfType(Radio).isRequired).isRequired,
};

export default RadioList;
