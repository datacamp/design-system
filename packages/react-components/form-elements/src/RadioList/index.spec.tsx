import '@testing-library/jest-dom/extend-expect';

import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import RadioList from './index';
import Radio from './Radio';

const testName = 'test-name';
const testValue = 'test value';
const testLabel = 'test label';

describe('<RadioList>', () => {
  it('renders the list', () => {
    const { getByLabelText, getByText } = render(
      <RadioList
        label={testLabel}
        name={testName}
        onChange={() => {}}
        value={testValue}
      >
        <Radio value={testValue}>item 1</Radio>
        <Radio value="value2">item 2</Radio>
      </RadioList>,
    );

    expect(getByText(testLabel) as HTMLElement).toBeInTheDocument();

    const item1 = getByLabelText('item 1') as HTMLElement;
    const item2 = getByLabelText('item 2') as HTMLElement;

    // all elements have the name
    expect(item1).toHaveAttribute('name', testName);
    expect(item2).toHaveAttribute('name', testName);
  });

  it('sets the dataAttributes on the checkbox element', () => {
    const { getByLabelText } = render(
      <RadioList
        label={testLabel}
        name={testName}
        onChange={() => {}}
        value={testValue}
      >
        <Radio dataAttributes={{ test: 'example' }} value="value1">
          radio 1
        </Radio>
        <Radio value="value2">radio 2</Radio>
      </RadioList>,
    );

    const radio1 = getByLabelText('radio 1', { selector: 'input' });

    expect(radio1).toHaveAttribute('data-test', 'example');
  });

  describe('required', () => {
    it('renders the indication "Required" above the radio group if required is passed as a prop', () => {
      const { getByText } = render(
        <RadioList
          label={testLabel}
          name={testName}
          onChange={() => {}}
          required
          value={testValue}
        >
          <Radio value={testValue}>item 1</Radio>
          <Radio value="value2">item 2</Radio>
        </RadioList>,
      );

      expect(getByText('Required')).toBeInTheDocument();
    });

    it('renders the indication "Optional" above the radio group if required=false is passed as a prop', () => {
      const { getByText } = render(
        <RadioList
          label={testLabel}
          name={testName}
          onChange={() => {}}
          required={false}
          value={testValue}
        >
          <Radio value={testValue}>item 1</Radio>
          <Radio value="value2">item 2</Radio>
        </RadioList>,
      );

      expect(getByText('Optional')).toBeInTheDocument();
    });
  });

  it('adds an error message under the radio group if error is passed as a prop', () => {
    const testError = 'Test error';

    const { getByText } = render(
      <RadioList
        errorMessage={testError}
        label={testLabel}
        name={testName}
        onChange={() => {}}
        value={testValue}
      >
        <Radio value={testValue}>item 1</Radio>
        <Radio value="value2">item 2</Radio>
      </RadioList>,
    );
    expect(getByText(testError)).toBeInTheDocument();
  });

  it('disables the whole component', () => {
    const { getByLabelText } = render(
      <RadioList
        disabled
        label="test-label"
        name={testName}
        onChange={() => {}}
        value="value1"
      >
        <Radio value="value1">item 1</Radio>
        <Radio value="value2">item 2</Radio>
      </RadioList>,
    );

    const radioElement1 = getByLabelText('item 1') as HTMLElement;
    const radioElement2 = getByLabelText('item 2') as HTMLElement;

    expect(radioElement1).toHaveAttribute('disabled');
    expect(radioElement2).toHaveAttribute('disabled');
  });

  it('disables an individual option', () => {
    const { getByLabelText } = render(
      <RadioList
        label="test-label"
        name={testName}
        onChange={() => {}}
        value="value1"
      >
        <Radio value="value1">item 1</Radio>
        <Radio disabled value="value2">
          item 2
        </Radio>
      </RadioList>,
    );

    const radioElement1 = getByLabelText('item 1') as HTMLElement;
    const radioElement2 = getByLabelText('item 2') as HTMLElement;

    expect(radioElement1).not.toHaveAttribute('disabled');
    expect(radioElement2).toHaveAttribute('disabled');
  });

  describe('interactivity', () => {
    it('has no selected option when value is invalid', () => {
      const { getByLabelText } = render(
        <RadioList
          label="test-label"
          name={testName}
          onChange={() => {}}
          value=""
        >
          <Radio value="value1">item 1</Radio>
          <Radio value="value2">item 2</Radio>
        </RadioList>,
      );

      const radioElement1 = getByLabelText('item 1') as HTMLElement;
      const radioElement2 = getByLabelText('item 2') as HTMLElement;

      expect(radioElement1).not.toHaveAttribute('checked');
      expect(radioElement2).not.toHaveAttribute('checked');
    });

    it('selects the correct option for the value', () => {
      const { getByLabelText } = render(
        <RadioList
          label="test-label"
          name={testName}
          onChange={() => {}}
          value="value1"
        >
          <Radio value="value1">item 1</Radio>
          <Radio value="value2">item 2</Radio>
        </RadioList>,
      );

      const radioElement1 = getByLabelText('item 1') as HTMLElement;
      const radioElement2 = getByLabelText('item 2') as HTMLElement;

      expect(radioElement1).toHaveAttribute('checked');
      expect(radioElement2).not.toHaveAttribute('checked');
    });

    it('calls onChange with the selected value', () => {
      const onChange = jest.fn();
      const { getByLabelText } = render(
        <RadioList
          label="test-label"
          name={testName}
          onChange={onChange}
          value="value1"
        >
          <Radio value="value1">item 1</Radio>
          <Radio value="value2">item 2</Radio>
        </RadioList>,
      );

      const radioElement2 = getByLabelText('item 2') as HTMLElement;

      userEvent.click(radioElement2);

      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange).toHaveBeenCalledWith('value2');
    });
  });

  describe('snapshots', () => {
    it(`renders a radio list`, async () => {
      const { container } = render(
        <RadioList
          errorMessage="test error"
          label={testLabel}
          name={testName}
          onChange={() => {}}
          value={testValue}
        >
          <Radio value={testValue}>item 1</Radio>
          <Radio value="value2">item 2</Radio>
          <Radio disabled value="value2">
            item 3
          </Radio>
        </RadioList>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
