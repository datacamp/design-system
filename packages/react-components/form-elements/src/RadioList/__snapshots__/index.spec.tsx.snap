// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioList> snapshots renders a radio list 1`] = `
.emotion-17 {
  border: 0;
  display: block;
  margin: 0;
  margin-top: 0;
  min-width: 0;
  padding: 0;
}

*:not(style) ~ .emotion-17 {
  margin-top: 16px;
}

.emotion-2 {
  display: block;
  margin-bottom: 12px;
  padding: 0;
  width: 100%;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5,25,45);
  font-family: Arial;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  color: #05192d;
  display: inline-block;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  font-weight: 800;
  margin-right: 4px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 20px;
  margin-top: 0;
  position: relative;
}

*:not(style) ~ .emotion-7 {
  margin-top: 10px;
}

.emotion-6 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5,25,45);
  font-family: Arial;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  margin-left: 24px;
}

.emotion-3 {
  opacity: 0;
  position: absolute;
  width: 0;
}

.emotion-3:active + span {
  border-color: #dd3400;
}

.emotion-3:focus + span {
  border-color: #dd3400;
}

.emotion-4 {
  border-color: #ff5400;
  border-radius: 50%;
  border-style: solid;
  border-width: 2px;
  box-sizing: border-box;
  height: 18px;
  opacity: 1;
  position: absolute;
  width: 18px;
}

.emotion-5 {
  background-color: #ff5400;
  border-radius: 50%;
  box-sizing: border-box;
  height: 8px;
  margin-left: 5px;
  opacity: 1;
  position: absolute;
  width: 8px;
}

.emotion-16 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5,25,45);
  font-family: Arial;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  color: #c01100;
  display: block;
  font-size: 14px;
  margin-top: 8px;
}

.emotion-14 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5,25,45);
  font-family: Arial;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  margin-left: 24px;
  opacity: 0.3;
}

.emotion-12 {
  opacity: 0;
  position: absolute;
  width: 0;
}

.emotion-13 {
  border-color: #ff5400;
  border-radius: 50%;
  border-style: solid;
  border-width: 2px;
  box-sizing: border-box;
  height: 18px;
  opacity: 0.3;
  position: absolute;
  width: 18px;
}

<div
  class="emotion-17"
>
  <span
    class="emotion-2"
  >
    <span
      class="emotion-1"
    >
      <span
        class="emotion-0"
      >
        test label
      </span>
    </span>
  </span>
  <div>
    <label
      class="emotion-7"
      for="test value"
    >
      <input
        checked=""
        class="emotion-3"
        id="test value"
        name="test-name"
        type="radio"
      />
      <span
        class="emotion-4"
      />
      <span
        class="emotion-5"
      />
      <span
        class="emotion-6"
      >
        item 1
      </span>
    </label>
    <label
      class="emotion-7"
      for="value2"
    >
      <input
        class="emotion-3"
        id="value2"
        name="test-name"
        type="radio"
      />
      <span
        class="emotion-4"
      />
      <span
        class="emotion-6"
      >
        item 2
      </span>
    </label>
    <label
      class="emotion-7"
      for="value2"
    >
      <input
        class="emotion-12"
        disabled=""
        id="value2"
        name="test-name"
        type="radio"
      />
      <span
        class="emotion-13"
      />
      <span
        class="emotion-14"
      >
        item 3
      </span>
    </label>
  </div>
  <span
    class="emotion-16"
  >
    test error
  </span>
</div>
`;
